app/code/Magento/Catalog/Block/Adminhtml/Category/Tab/Product.php

```diff
@@ -29,20 +32,37 @@ class Product extends \Magento\Backend\Block\Widget\Grid\Extended
      */
     protected $_productFactory;
 
+    /**
+     * @var Status
+     */
+    protected $_status;
+
+    /**
+     * @var Visibility
+     */
+    protected $_visibility;
+
     /**
      * @param \Magento\Backend\Block\Template\Context $context
      * @param \Magento\Backend\Helper\Data $backendHelper
      * @param \Magento\Catalog\Model\ProductFactory $productFactory
      * @param \Magento\Framework\Registry $coreRegistry
      * @param array $data
+     * @param Visibility|null $visibility
+     * @param Status|null $status
      */
     public function __construct(
         \Magento\Backend\Block\Template\Context $context,
         \Magento\Backend\Helper\Data $backendHelper,
         \Magento\Catalog\Model\ProductFactory $productFactory,
         \Magento\Framework\Registry $coreRegistry,
-        array $data = []
+        array $data = [],
+        Visibility $visibility = null,
+        Status $status = null
     ) {
+        $this->_visibility = $visibility ?: ObjectManager::getInstance()->get(Visibility::class);
+        $this->_status = $status ?: ObjectManager::getInstance()->get(Status::class);
+
```

Comment:
It would be nice to rearrange class variable assignations in the same order as defined in constructor, although it's not strictly needed

Reply:
Done